agent/src/main/java/com/teamscale/jacoco/agent/commit_resolution/git_properties/GitPropertiesLocatorUtils.java

```diff
@@ -55,22 +59,50 @@ public static String getRevisionFromGitProperties(File jarFile) throws IOExcepti
 	 * @throws URISyntaxException under certain circumstances if parsing the URL fails. This should be treated the same
 	 *                            as a null search result but the exception is preserved so it can be logged.
 	 */
-	public static File extractGitPropertiesSearchRoot(URL jarOrClassFolderUrl) throws URISyntaxException {
+	public static Pair<File, Boolean> extractGitPropertiesSearchRoot(
+			URL jarOrClassFolderUrl) throws URISyntaxException, IOException, NoSuchMethodException,
+			IllegalAccessException, InvocationTargetException {
 		String protocol = jarOrClassFolderUrl.getProtocol().toLowerCase();
-		if (protocol.equals("file") && org.conqat.lib.commons.string.StringUtils.endsWithOneOf(
-				jarOrClassFolderUrl.getPath().toLowerCase(), ".jar", ".war", ".ear", ".aar")) {
-			return new File(jarOrClassFolderUrl.toURI());
-		} else if (protocol.equals("jar")) {
-			// used e.g. by Spring Boot. Example: jar:file:/home/k/demo.jar!/BOOT-INF/classes!/
-			Matcher matcher = JAR_URL_REGEX.matcher(jarOrClassFolderUrl.toString());
-			if (!matcher.matches()) {
+		switch (protocol) {
+			case "file":
+				if (org.conqat.lib.commons.string.StringUtils.endsWithOneOf(
+						jarOrClassFolderUrl.getPath().toLowerCase(), ".jar", ".war", ".ear", ".aar")) {
+					return Pair.createPair(new File(jarOrClassFolderUrl.toURI()), true);
+				}
+				break;
+			case "jar":
+				// used e.g. by Spring Boot. Example: jar:file:/home/k/demo.jar!/BOOT-INF/classes!/
+				Matcher matcher = JAR_URL_REGEX.matcher(jarOrClassFolderUrl.toString());
+				if (!matcher.matches()) {
+					return null;
+				}
+				return Pair.createPair(new File(matcher.group(1)), true);
+			case "vfs":
+				return getVfsContentFolder(jarOrClassFolderUrl);
+			default:
 				return null;
-			}
-			return new File(matcher.group(1));
 		}
 		return null;
 	}
 
+	private static Pair<File, Boolean> getVfsContentFolder(
+			URL jarOrClassFolderUrl) throws IOException, NoSuchMethodException, IllegalAccessException, InvocationTargetException {
+		// used by JBoss EAP and Wildfly. Example: vfs:/content/helloworld.war/WEB-INF/classes
+		Object virtualFile = jarOrClassFolderUrl.openConnection().getContent();
+		Class<?> virtualFileClass = virtualFile.getClass();
+		// obtain the physical location of the class file. It is created on demand in <jboss-installation-dir>/standalone/tmp/vfs
+		Method getPhysicalFileMethod = virtualFileClass.getMethod("getPhysicalFile");
+		File file = (File) getPhysicalFileMethod.invoke(virtualFile);
+		if (file.getParentFile() == null || file.getParentFile().getParentFile() == null) {
+			return null;
+		}
+		// the physical location of the class is in, e.g., WEB-INF/classes folder.
```

Comment:
I don't think this will work in all cases. the URLs I've seen from our customer did not look like this but were at a deeper nesting level. I think we instead have to manipulate the URL so it points to `helloworld.war` and then resolve that. I.e. extract the first war/jar/ear file from the url, throw away the rest of the url path and resolve the jar file to disk

Reply:
Ok makes sense, thanks.